docs/performance.rst

```diff
@@ -0,0 +1,139 @@
+-------------------------------
+
+After achieving good performance with **NumPy**, consider trying **Numba** for further improvements.
+
+Even if you're new to Numba, **ChatGPT** can generate optimized solutions easily. Keep in mind:
+
+- **Numba moves back to for-loops** (unlike NumPy, which prefers vectorization).
+- GPT's first solution may use ``range`` instead of ``prange``. Prompt it with:
+
+  .. code-block:: text
+
+     "Can we parallelize this with prange?"
+
+- If your code involves **common counters**, **atomic operations** may become a bottleneck.
+  Ask GPT about:
+
+  .. code-block:: text
+
+     "Can we use thread-local storage to avoid atomic operations?"
+
+- **Numba may be slower than NumPy for small arrays** (e.g., thousands or tens of thousands of elements). Test with **at least 1 million elements**.
+
+Further Tricks
+^^^^^^^^^^^^^^
+
+- Don't duplicate: Sometimes reporting will duplicate transmission code and need to be combined.
+- Never append. There may be cases where you are collecting information as it happens without knowing ahead of time how many rows/entries/elements you'll need. This is easy in Python using list appending, for example, but that's a performance killer. Really try to find a way to figure out ahead of time how many entries there will be, and then allocate memory for that, and insert into the existing row.
+- Some components have long time-scales, like mortality. By default you are probably going to end up do most component steps every timestep. You can probably get away with doing mortality updates, for example, far less often. You can experiement with weekly, fortnightly or monthly updates, depending on the timescale of the component you're optimizing. Just be sure to move everything forward by a week if you're only doing the update every week. And expect "blocky" plots. Note that there are fancier solutions like 'strided sharding' (details omitted).
+
+When **prompting AI**, use **questions rather than directives**. Example:
+
+.. code-block:: text
+
+   "Do you think it might be better to...?"
+
+This prevents oversteering the AI into suboptimal solutions.
+
+Beyond Numba: C and OpenMP
+--------------------------
+
+If the best **Numba** solution still isn't fast enough, consider **compiled C**.
+
+- Use **ctypes** to call C functions from Python.
+- Mention **"use OpenMP"** in AI prompts if parallelization is possible.
+- Ask:
+
+  .. code-block:: text
+
+     "Can you generate an OpenMP solution with the best pragmas?"
+
+- The more CPU cores available, the **greater the potential speedup**. That said, it's usually a case of diminishing returns as one goes from 8 cores to 16 and to 32. Our research shows that often you're better off running 4 sims across 8 cores each than running 1 sim on all 32 cores available. Also be aware that with both Numba and OpenMP, you can constrain the number of cores used to less than the number available but setting the appropriate environment variable.
```

Comment:
"the number available **_by_** setting the appropriate environment variable."

Numba environment variable - `NUMBA_NUM_THREADS`.
OpenMP environment variable - `OMP_NUM_THREADS`.